org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend

```diff
@@ -2703,6 +2753,210 @@ class MarkedString {
+	 * complementary information to the `title`.
+	 *
+	 * Examples: "3/25 files", "project/src/module2", "node_modules/some_dep".
+	 * If unset, the previous progress message (if any) is still valid.
+	 */
+	String message
+
+	/**
+	 * Optional progress percentage to display (value 100 is considered 100%).
+	 * If not provided infinite progress is assumed and clients are allowed
+	 * to ignore the `percentage` value in subsequent in report notifications.
+	 *
+	 * The value should be steadily rising. Clients are free to ignore values
+	 * that are not following this rule.
+	 */
+	Integer percentage;
+}
+
+/**
+ *  The notification payload about progress reporting.
+ *
+ * Since 3.15.0
+ */
+@JsonRpcData
+class WorkDoneProgressReport implements WorkDoneProgressNotification {
+
+	override WorkDoneProgressKind getKind() {
+		WorkDoneProgressKind.report
+	}
+
+	/**
+	 * Controls enablement state of a cancel button. This property is only valid if a cancel
+	 * button got requested in the `WorkDoneProgressStart` payload.
+	 *
+	 * Clients that don't support cancellation or don't support control the button's
+	 * enablement state are allowed to ignore the setting.
+	 */
+	Boolean cancellable
+
+	/**
+	 * Optional, more detailed associated progress message. Contains
+	 * complementary information to the `title`.
+	 *
+	 * Examples: "3/25 files", "project/src/module2", "node_modules/some_dep".
+	 * If unset, the previous progress message (if any) is still valid.
+	 */
+	String message
+
+	/**
+	 * Optional progress percentage to display (value 100 is considered 100%).
+	 * If not provided infinite progress is assumed and clients are allowed
+	 * to ignore the `percentage` value in subsequent in report notifications.
+	 *
+	 * The value should be steadily rising. Clients are free to ignore values
+	 * that are not following this rule.
+	 */
+	Integer percentage;
+}
+
+/**
+ *  The notification payload about progress reporting.
+ * Signaling the end of a progress reporting is done using the following payload:
+ * 
+ * Since 3.15.0
+ */
+@JsonRpcData
+class WorkDoneProgressEnd implements WorkDoneProgressNotification {
+
+	override WorkDoneProgressKind getKind() {
+		WorkDoneProgressKind.end
+	}
+
+	/**
+	 * Optional, a final message indicating to for example indicate the outcome
+	 * of the operation.
+	 */
+	String message
+}
+
+/**
+ *  The base protocol offers also support to report progress in a generic fashion. 
+ *  This mechanism can be used to report any kind of progress including work done progress 
+ *  (usually used to report progress in the user interface using a progress bar) 
+ *  and partial result progress to support streaming of results. 
+ *  A progress notification has the following properties:
+ * 
+ * Since 3.15.0
+ */
+@JsonRpcData
+class ProgressParams {
+
+	Either<String, Number> token
+
+	@JsonAdapter(WorkDoneProgressNotificationAdapter.Factory)
+	WorkDoneProgressNotification value;
```

Comment:
Not sure if this was already discussed:
```
interface ProgressParams<T> {
	/**
	 * The progress token provided by the client or server.
	 */
	token: ProgressToken;

	/**
	 * The progress data.
	 */
	value: T;
}
```

What is the purpose of the type parameter `T` in the spec and are we missing an opportunity / a valid use case by typing this to `WorkDoneProgressNotification`?

Reply:
It's strange that they didn't constrain `T` to any supertype. But Section [Work Done Progress](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workDoneProgress) states

> The value payload of a work done progress notification can be of three different forms.

These "different forms" are exactly the three implementations of the WorkDoneProgressNotification interface. I think this is a sufficient approximation of what they are trying to express in the spec.

Reply:
And in fact generics will add many <?> in the code. I am not sur it's a good idea.